### PR TITLE
Fixing PropertyGrid memory leaks

### LINKED ISSUE #6943: PropertyGrid accessible objects are leaking in memory
### .NET version

7.0

### Did it work in .NET Framework?

No

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No

### Issue description

`PropertyGrid` accessible objects are left in memory after the `PropertyGrid` control instance was inspected with accessibility tool (like `Accessibility Insights` or `Inspect`) and the form with `PropertyGrid` is closed, garbage is collected, but accessibility tool is still opened.

Report from WinDbg:
image

**Expected behavior:**
Amount of the leaked objects should be minimized to the uncleanable static objects only.

### Steps to reproduce

1. Create a parent form with two buttons: one will be opening a child form with the `PropertyGrid` (`SelectedObject` set to `this`) control and the other will collect the garbage.
2. Press the second button to open the child form with `PropertyGrid`.
3. Start the `Accessibility Insights` tools.
4. Run automated checks for the `PropertyGrid` control using AI.
5. Close the child form.
6. Clear memory with the first button of the parent form.
7. DON'T close the AI tool.
8. Look for `PropertyGrid` accessible objects left in memory (using `WinDbg`, or `dotMemory`, or another suitable tool).

### PR DESCRIPTION
**Pull request by ArtemTatarinov**
Fixes #6943

## Proposed changes
- Amount of leaked objects decreased from 595 to 39
- Use the `UiaDisconnectProvider` method through the hierarchy of the classes to free objects held by accessibility tool
- Only static comparers should be left at memory as amount of memory allocated for them isn't increasing after several recurrences of the repro steps described in issue

**This PR requires further development:**
- 39 objects are still left in memory, some of them are accessible objects. They should be tried to clean, except for the Comparers/Sorters which are most possibly are static system objects.
- Quite possible that nullifying `_gridView` and other objects won't be necessary when all provider will be disconnected property
- I've met a crash at the code line `Debug.Assert(result == 0)` for the result of disconnecting the third button of the `_viewSortButtons` collection if debugger is detached, and E_FAIL result if attached - it requires further investigation
- There was a crash scenario at NetFx:  start AI, run internal test app, click "Dialogs" button, click "Print dialog" button, then "Folder browser dialog" - disconnecting `gridEntry.AccessibilityObject` at the `DisconnectChildren` method at the `PropertyGridView` class caused NRE. I couldn't reproduce this issue at the Core branch, but it's better to double check it.

## Customer Impact
- PropertyGrid memory will stop leaking

## Regression? 
- No

## Risk
- Minimal

### Before
image

### After
image

## Test methodology
- Manual testing (_using WinDbg, see the screenshots above_)
- CTI team

## Test environment(s)
- Microsoft Windows [Version 10.0.19043.1586]
- .NET 7.0.0-preview.3.22159.12

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/6946)

### CODE REVIEW

**Comment by Tanya-Solyanik** [comment]
@vladimir-krestov  - please prioritize this work because we introduced these leak when adding UIA providers

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs [hunk 1]
```diff
@@ -4255,6 +4258,54 @@ protected unsafe override void WndProc(ref Message m)
 
                     m.ResultInternal = 1;
                     return;
+                case User32.WM.DESTROY:
+                    base.WndProc(ref m);
+                    if (OsVersion.IsWindows8OrGreater && IsAccessibilityObjectCreated)
+                    {
+                        _gridView?.DisconnectChildren();
+                        HRESULT result = UiaCore.UiaDisconnectProvider(_gridView?.AccessibilityObject);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs:4266]: Do not call `UiaDisconnectProvide` if _gridView is null or _gridView does not have accessibility object created. And the same applies to all other subcontrols.

File: src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs [hunk 2]
```diff
@@ -4255,6 +4258,54 @@ protected unsafe override void WndProc(ref Message m)
...
+                        Debug.Assert(result == 0);
+                        result = UiaCore.UiaDisconnectProvider(_helpPane?.AccessibilityObject);
+                        Debug.Assert(result == 0);
+                        result = UiaCore.UiaDisconnectProvider(_commandsPane?.AccessibilityObject);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs:4270]: Are we really creating providers for all inner controls?

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs:4270]: Please use `HRESULT.S_OK` if possible

Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs:4270]: We should use these extensions
https://github.com/dotnet/winforms/blob/5d454f4bcee2bf6eced54624720f2ddf0980e9c8/src/System.Windows.Forms.Primitives/src/Interop/Interop.HRESULT.cs#L83-L92

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/PropertyGrid.cs:4270]: We want to Debug.Assert on S_FALSE

**Comment by dmitrii-drobotov** [comment]
We'll create a new PR once we get around to it and use this one as a starting point.

**Pull request closed without merging**